src/common/survival_util.h

```diff
@@ -8,15 +8,68 @@
 #ifndef XGBOOST_COMMON_SURVIVAL_UTIL_H_
 #define XGBOOST_COMMON_SURVIVAL_UTIL_H_
 
+/*
+- Formulas are motivated from document -
+  http://members.cbio.mines-paristech.fr/~thocking/survival.pdf
+- Detailed Derivation of Loss/Gradient/Hessian -
+  https://github.com/avinashbarnwal/GSOC-2019/blob/master/doc/Accelerated_Failure_Time.pdf
+*/
+
 #include <xgboost/parameter.h>
 #include <memory>
+#include <algorithm>
 #include "probability_distribution.h"
 
 DECLARE_FIELD_ENUM_CLASS(xgboost::common::ProbabilityDistributionType);
 
 namespace xgboost {
 namespace common {
 
+#ifndef __CUDACC__
+
+using std::log;
+using std::fmax;
+
+#endif  // __CUDACC__
+
+enum class CensoringType : uint8_t {
+  kUncensored, kRightCensored, kLeftCensored, kIntervalCensored
+};
+
+namespace aft {
+
+// Allowable range for gradient and hessian. Used for regularization
+constexpr double kMinGradient = -15.0;
+constexpr double kMaxGradient = 15.0;
+constexpr double kMinHessian = 1e-16;  // Ensure that no data point gets zero hessian
+constexpr double kMaxHessian = 15.0;
+
+constexpr double kEps = 1e-12;  // A denomitor in a fraction should not be too small
+
+// Clip (limit) x to fit range [x_min, x_max].
+// If x < x_min, return x_min; if x > x_max, return x_max; if x_min <= x <= x_max, return x.
+// This function assumes x_min < x_max; behavior is undefined if this assumption does not hold.
+XGBOOST_DEVICE
+inline double Clip(double x, double x_min, double x_max) {
+  if (x < x_min) {
+    return x_min;
+  }
+  if (x > x_max) {
+    return x_max;
+  }
+  return x;
+}
+
+template<typename Distribution>
+XGBOOST_DEVICE inline double
+GetLimitGradAtInfPred(CensoringType censor_type, bool sign, double sigma);
```

Comment:
@trivialfis This line is a forward declaration. The definition (with template specialization) is at the bottom of this source file.

Reply:
@hcho3 Yes, but I'm not sure we do we need to split them.

Reply:
@trivialfis This is also because I define template specializations `GetLimitGradAtInfPred<NormalDistribution>`, `GetLimitGradAtInfPred<LogisticDistribution>`, and `GetLimitGradAtInfPred<ExtremeDistribution>`. The base template `GetLimitGradAtInfPred<T>` lacks a definition because we cannot define it unless `T` is a known distribution type.

Reply:
@hcho3  Got it.  Thanks for detailed explanation.